doc/fgpu.design.rst

```diff
@@ -259,14 +339,121 @@ written as :math:`pm + s`. Let :math:`z^r` denote the array :math:`z_r, z_{n+r},
+
+The fourth array when :math:`r=3` (i.e. :math:`z^{3}`)
+
+========================  ========
+   Inputs                 Index(i)
+------------------------  --------
+  qn + r                     i
+========================  ========
+:math:`0 \cdot 4 + 3`        3
+:math:`1 \cdot 4 + 3`        7
+:math:`2 \cdot 4 + 3`        11
+...               ...
+...               ...
+:math:`8191 \cdot 4 + 3`   32767
+========================  ========
+
+What this shows is that each sub-array consists of samples from the initial array :math:`z` indexed
+by :math:`i=qn+r` where each sample is every :math:`4^{th}` and offset by :math:`r`. Pictorially
+this looks like,
+
+.. image:: images/z_array.png
+   :width: 600
+(click image to enlarge)
+
+Right, so we have separate sub-arrays as indexed from the initial array, what happens next? These various
+:math:`z^{r}` arrays are fed to cuFFT yielding 4 complex-to-complex transforms. These separate transforms
+now need to be combined to form a single real-to-complex transform of the full initial size. An inconvenience
+of this structure is that :math:`z^r` is not a contiguous set of input samples, but a strided array.
+While cuFFT does support both strided inputs and batched transformations, we cannot batch over :math:`r`
+and over multiple spectra at the same time as it only supports a single batch dimension with corresponding
+stride. We solve this by modifying the PFB kernel to reorder its output such that each :math:`z^r` is output
+contiguously. This can be done by shuffling some bits in the output index (because we assume powers of two
+everywhere).
+
+To see how the :math:`k` indexing works, :math:`k = pm + s` and is dealt with in a similar manner as above.
+Parameter :math:`m = 8192` (in this example), and :math:`p` has a range :math:`0` to :math:`n-1`
+(i.e. :math:`p = 0` to :math:`p = 3` as :math:`n = 4` in our example); and :math:`s` takes on the
+range :math:`0` to :math:`m-1` (i.e. :math:`s = 0` to :math:`s = 8191`).
+
+Looking at this practically,
+
+When :math:`p = 0`
+
+===========================  =========
+   Inputs                     Index(k)
+---------------------------  ---------
+  pm + s                      k
+===========================  =========
+:math:`0 \cdot 8192` + 0      0
+:math:`0 \cdot 8192` + 1      1
+:math:`0 \cdot 8192` + 2      2
+...               ...
+...               ...
+:math:`0 \cdot 8192` + 8191   8191
+===========================  =========
+
+... (intentionally skipping a few to conserve space)
+
+When :math:`p=3`
+
+===========================  ========
+   Inputs                    Index(k)
+---------------------------  --------
+  pm + s                         k
+===========================  ========
+:math:`3 \cdot 8192` + 0      24576
+:math:`3 \cdot 8192` + 1      24577
+:math:`3 \cdot 8192` + 2      24578
+...              ...
+...              ...
+:math:`3 \cdot 8192` + 8191   32767
+===========================  ========
```

Comment:
Again, these tables take a huge amount of space and could be replaced by a couple of sentences.